### PR TITLE
PLFM 5709: Send download/upload logs to firehose

### PR DESCRIPTION
**Pull request by marcomarasca**

### CODE REVIEW

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/statistics/StatisticsEventsCollector.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	 */
+	<E extends StatisticsEvent> void collectEvent(E event);
+
+	/**
+	 * Accepts the given list of {@link StatisticsEvent}s in order to collect statistics about the batch of events. If
+	 * this call is invoked within a transaction if will be called only after the transaction is committed.
+	 * 
+	 * @param events A batch of events to collect statistics about
+	 */
+	<E extends StatisticsEvent> void collectEvents(List<E> events);
```
Comment (marcomarasca) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/statistics/StatisticsEventsCollector.java:29]: Think about having a queue to batch and a background job to push the events to firehose

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/statistics/StatisticsEventsCollectorImpl.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+		afterCommit(() -> log(events));
+	}
+
+	private void afterCommit(Action action) {
+		
+		if (transactionSynchronization.isActualTransactionActive()) {
+			transactionSynchronization.registerSynchronization(new TransactionSynchronization() {
+				@Override
+				public void afterCommit() {
+					action.proceed();
```
Comment (marcomarasca) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/statistics/StatisticsEventsCollectorImpl.java]: Queue the events here

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/statistics/records/StatisticsFileEventLogRecord.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	private Long projectId;
+	private String fileHandleId;
+	private FileHandleAssociateType associateType;
+	private String associateId;
+	private String stack;
+	private String instance;
+
+	@JsonIgnore
+	@Override
+	public byte[] toBytes() {
```
Comment (marcomarasca) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/statistics/records/StatisticsFileEventLogRecord.java]: Default method or extract?

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/statistics/records/StatisticsEventMock.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package org.sagebionetworks.repo.manager.statistics.records;
+
+import org.sagebionetworks.repo.manager.statistics.events.StatisticsEvent;
+
+public class StatisticsEventMock implements StatisticsEvent {
```
Comment (marcomarasca) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/statistics/records/StatisticsEventMock.java]: Stub

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/statistics/records/StatisticsFileEventRecordProviderUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	public void testGetRecordForNonExistingProject() {
+		Long userId = 123L;
+		String fileHandleId = "123";
+		String associateId = "123";
+		FileHandleAssociateType associateType = FileHandleAssociateType.FileEntity;
+
+		StatisticsFileEvent event = new StatisticsFileEvent(StatisticsFileActionType.FILE_DOWNLOAD, userId,
+				fileHandleId, associateId, associateType);
+		
+		when(mockProjectResolver.resolveProject(associateType, associateId)).thenThrow(NotFoundException.class);
```
Comment (marcomarasca) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/statistics/records/StatisticsFileEventRecordProviderUnitTest.java]: Spy on the logger

File: services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/statistics/records/StatisticsLogRecordProviderFactoryUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.mockito.junit.jupiter.MockitoSettings;
+import org.mockito.quality.Strictness;
+import org.sagebionetworks.repo.manager.statistics.events.StatisticsEvent;
+
+@ExtendWith(MockitoExtension.class)
+@MockitoSettings(strictness = Strictness.LENIENT)
```
Comment (marcomarasca) [inline services/repository-managers/src/test/java/org/sagebionetworks/repo/manager/statistics/records/StatisticsLogRecordProviderFactoryUnitTest.java]: what is this?

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/file/FileHandleManagerImpl.java [hunk 7]
```diff
@@ -424,6 +437,10 @@ private String getUrlForS3FileHandle(S3FileHandle handle) {
 	private String getUrlForGoogleCloudFileHandle(GoogleCloudFileHandle handle) {
 		return googleCloudStorageClient.createSignedUrl(handle.getBucketName(), handle.getKey(), (int) PRESIGNED_URL_EXPIRE_TIME_MS, com.google.cloud.storage.HttpMethod.GET).toExternalForm();
 	}
+	
+	private StatisticsFileEvent buildFileDownloadEvent(Long userId, FileHandleAssociation association) {
```
Comment (marcomarasca) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/file/FileHandleManagerImpl.java]: Extract this into a utility with a static method

File: services/repository/src/main/java/org/sagebionetworks/repo/web/service/EntityServiceImpl.java [hunk 8]
```diff
@@ -236,6 +242,9 @@ private void fireAfterCreateEntityEvent(UserInfo userInfo, Entity entity, Entity
 				}
 			}
 		}
+		if (EntityType.file.equals(type) && entity instanceof FileEntity) {
```
Comment (marcomarasca) [inline services/repository/src/main/java/org/sagebionetworks/repo/web/service/EntityServiceImpl.java]: Put this logic into the FileEntityMetadataProvider

File: services/repository/src/main/java/org/sagebionetworks/repo/web/service/EntityServiceImpl.java [hunk 9]
```diff
@@ -258,6 +267,14 @@ private void fireAfterUpdateEntityEvent(UserInfo userInfo, Entity entity, Entity
 				}
 			}
 		}
+		if (EntityType.file.equals(type) && entity instanceof FileEntity && wasNewVersionCreated) {
```
Comment (marcomarasca) [inline services/repository/src/main/java/org/sagebionetworks/repo/web/service/EntityServiceImpl.java]: FileEntityMetadataProvider

File: services/repository/src/main/java/org/sagebionetworks/repo/web/service/EntityServiceImpl.java [hunk 10]
```diff
@@ -258,6 +267,14 @@ private void fireAfterUpdateEntityEvent(UserInfo userInfo, Entity entity, Entity
...
+			sendFileUploadEvent(userInfo.getId(), (FileEntity) entity);
+		}
+	}
+	
+	private void sendFileUploadEvent(Long userId, FileEntity entity) {
+		StatisticsFileEvent event = new StatisticsFileEvent(StatisticsFileActionType.FILE_UPLOAD, userId, entity.getDataFileHandleId(), entity.getId(), FileHandleAssociateType.FileEntity);
+		statisticsCollector.collectEvent(event);
```
Comment (marcomarasca) [inline services/repository/src/main/java/org/sagebionetworks/repo/web/service/EntityServiceImpl.java]: Use static method

File: services/workers/src/main/java/org/sagebionetworks/file/worker/BulkFileDownloadWorker.java [hunk 11]
```diff
@@ -255,6 +263,20 @@ public String writeOneFileToZip(ZipOutputStream zipOut, long zipFileSize,
 			downloadTemp.delete();
 		}
 	}
+	
+	private void collectDownloadStatistics(Long userId, List<FileDownloadSummary> results) {
+		
+		List<StatisticsFileEvent> downloadEvents = results.stream()
+			// Only collects stats for successful summaries
+			.filter(summary -> FileDownloadStatus.SUCCESS.equals(summary.getStatus()))
+			.map(summary -> new StatisticsFileEvent(StatisticsFileActionType.FILE_DOWNLOAD, userId, summary.getFileHandleId(), summary.getAssociateObjectId(), summary.getAssociateObjectType()))
```
Comment (marcomarasca) [inline services/workers/src/main/java/org/sagebionetworks/file/worker/BulkFileDownloadWorker.java]: Static method utility

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableEntityManagerImpl.java [hunk 12]
```diff
@@ -358,9 +365,23 @@ RowReferenceSet appendRowsToTable(UserInfo user, List<ColumnModel> columns,
 			refs.add(ref);
 		}
 		results.setRows(refs);
+		sendFileUploadEvents(user.getId(), delta);
```
Comment (marcomarasca) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableEntityManagerImpl.java]: Exclude files that are already in the table: getFileHandleIdsAssociatedWithTable

File: services/repository/src/test/java/org/sagebionetworks/repo/web/EntityServiceImplUnitTest.java [hunk 13]
```diff
@@ -153,5 +154,28 @@ public void testFireUpdateTriggersNewVersion() {
...
+	
+	@Test
+	public void testFireCreateWithCollectStatistics() {
+		FileEntity fileEntity = new FileEntity();
+		
+		fileEntity.setId("123");
+		fileEntity.setDataFileHandleId("456");
+		
+		entityService.createEntity(userInfo.getId(), fileEntity, null);		
+	}
```
Comment (marcomarasca) [inline services/repository/src/test/java/org/sagebionetworks/repo/web/EntityServiceImplUnitTest.java:166]: Remove old code

File: services/repository/src/test/java/org/sagebionetworks/repo/web/EntityServiceImplUnitTest.java [hunk 14]
```diff
@@ -153,5 +154,28 @@ public void testFireUpdateTriggersNewVersion() {
...
+		FileEntity fileEntity = new FileEntity();
+		
+		fileEntity.setId("123");
+		fileEntity.setDataFileHandleId("456");
+		
+		when(mockEntityManager.updateEntity(userInfo, fileEntity, false, null)).thenReturn(true);
+		
+		entityService.updateEntity(userInfo.getId(), fileEntity, false, null);
+		
+	}
```
Comment (marcomarasca) [inline services/repository/src/test/java/org/sagebionetworks/repo/web/EntityServiceImplUnitTest.java:179]: Remove old code

File: services/repository/src/test/java/org/sagebionetworks/repo/web/service/metadata/FileEntityMetadataProviderTest.java [hunk 15]
```diff
@@ -30,9 +37,6 @@
 	public void before(){
 		MockitoAnnotations.initMocks(this);
```
Comment (marcomarasca) [inline services/repository/src/test/java/org/sagebionetworks/repo/web/service/metadata/FileEntityMetadataProviderTest.java:38]: We might not need this

File: services/repository/src/test/java/org/sagebionetworks/repo/web/service/metadata/FileEntityMetadataProviderTest.java [hunk 16]
```diff
@@ -9,19 +12,23 @@
...
 import org.mockito.MockitoAnnotations;
-import org.sagebionetworks.repo.manager.EntityManager;
+import org.sagebionetworks.repo.manager.statistics.StatisticsEventsCollector;
+import org.sagebionetworks.repo.manager.statistics.events.StatisticsFileEvent;
 import org.sagebionetworks.repo.model.EntityHeader;
 import org.sagebionetworks.repo.model.FileEntity;
 import org.sagebionetworks.repo.model.UserInfo;
-import org.springframework.test.util.ReflectionTestUtils;
 
 public class FileEntityMetadataProviderTest  {
```
Comment (marcomarasca) [inline services/repository/src/test/java/org/sagebionetworks/repo/web/service/metadata/FileEntityMetadataProviderTest.java:24]: Annotate with mockito extension

File: services/repository/src/test/java/org/sagebionetworks/repo/web/service/metadata/FileEntityMetadataProviderTest.java [hunk 17]
```diff
@@ -55,49 +59,65 @@ public void before(){
...
+	public void testEntityCreated() {
+		fileEntity.setDataFileHandleId("1");
+		provider.entityCreated(userInfo, fileEntity);
+		verify(mockStatisticsCollector, times(1)).collectEvent(any(StatisticsFileEvent.class));
+	}
+	
+	@Test
+	public void testEntityUpdatedWithNewVersion() {
+		fileEntity.setDataFileHandleId("1");
+		provider.entityUpdated(userInfo, fileEntity, true);
```
Comment (marcomarasca) [inline services/repository/src/test/java/org/sagebionetworks/repo/web/service/metadata/FileEntityMetadataProviderTest.java:114]: Extract to local variable

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableEntityManagerImpl.java [hunk 18]
```diff
@@ -358,9 +368,35 @@ RowReferenceSet appendRowsToTable(UserInfo user, List<ColumnModel> columns,
...
+			StatisticsFileEventUtils.buildFileUploadEvent(userId, fileHandleId.toString(), tableId, FileHandleAssociateType.TableEntity)
+		).collect(Collectors.toList());
+		
+		if (!downloadEvents.isEmpty()) {
+			statisticsCollector.collectEvents(downloadEvents);
+		}
+		
+	}
+	
+	Set<Long> getFileHandleIdsNotAssociatedWithTable(String tableId, Set<Long> fileHandleIds) {
```
Comment (marcomarasca) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/table/TableEntityManagerImpl.java:390]: Add comment

File: services/workers/src/main/resources/main-scheduler-spb.xml [hunk 19]
```diff
@@ -28,6 +28,7 @@
 	<!-- non worker triggers -->
 	<import resource="classpath:managers-topic-publisher-triggers-import.xml" />
 	<import resource="classpath:cloudwatch-trigger-spb.xml" />
+	<import resource="classpath:managers-triggers-import.xml" />
```
Comment (marcomarasca) [inline services/workers/src/main/resources/main-scheduler-spb.xml:31]: Create a dedicated config for triggers that come from the repo services (including the memory-logger-trigger-spb)

**Review by john-hill — APPROVED** [review]

**Pull request merged**